src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java

```diff
@@ -127,111 +105,210 @@ private void rcptToCmd() {
     if (email.getBcc() != null) {
       recipientAddrs.addAll(email.getBcc());
     }
-    rcptToCmd(recipientAddrs, 0);
+    return recipientAddrs.stream().map(r -> {
+      final String email;
+      if (EmailAddress.POSTMASTER.equalsIgnoreCase(r)) {
+        email = r;
+      } else {
+        email = new EmailAddress(r).getEmail();
+      }
+      return email;
+    }).collect(Collectors.toList());
   }
 
-  private void rcptToCmd(List<String> recipientAddrs, int i) {
+  private Future<Boolean> sendMailEvenlope() {
+    Promise<Boolean> evenlopePromise = Promise.promise();
     try {
-      final String recipientAddr = recipientAddrs.get(i);
-      final String email;
-      if (EmailAddress.POSTMASTER.equalsIgnoreCase(recipientAddr)) {
-        email = recipientAddr;
+      if (checkSize()) {
+        final String mailFromLine = "MAIL FROM:<" + mailFromAddress() + ">" + sizeParameter();
+        final List<String> allRecipients = allRecipients();
+        if (config.isPipelining() && connection.getCapa().isCapaPipelining()) {
+          final List<String> groupCommands = new ArrayList<>();
+          groupCommands.add(mailFromLine);
+          groupCommands.addAll(allRecipients.stream().map(r -> "RCPT TO:<" + r + ">").collect(Collectors.toList()));
+          groupCommands.add("DATA");
+          connection.writeCommands(groupCommands, evenlopeResult -> {
+            if (groupCommands.size() != evenlopeResult.size()) {
+              evenlopePromise.fail("Sent " + groupCommands.size() + " commands, but got " + evenlopeResult.size() + " responses.");
+            } else {
+              // result follows the same order in the commands list
+              for (int i = 0; i < evenlopeResult.size(); i ++) {
+                String message = evenlopeResult.get(i);
+                if (log.isDebugEnabled()) {
+                  log.debug("S: " + message);
```

Comment:
don't debug data, it allows log injection

Reply:
Done